internal/services/notification/notification_test.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 Sorint.lab
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package notification
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/rs/zerolog"
+	"github.com/sorintlab/errors"
+
+	"agola.io/agola/internal/services/config"
+	"agola.io/agola/internal/sqlg/sql"
+	"agola.io/agola/internal/testutil"
+	"agola.io/agola/services/notification/types"
+)
+
+func setupNotificationService(pctx context.Context, t *testing.T, log zerolog.Logger, dir string) *testNotificationService {
+	ctx, cancel := context.WithCancel(pctx)
+
+	dbType := testutil.DBType(t)
+	_, _, dbConnString := testutil.CreateDB(t, log, ctx, dir)
+
+	c := config.Config{
+		Notification: config.Notification{
+			DB: config.DB{
+				Type:       dbType,
+				ConnString: dbConnString,
+			},
+		},
+	}
+
+	ns, err := NewNotificationService(ctx, log, &c)
+	if err != nil {
+		cancel()
+		t.Fatalf("unexpected err: %v", err)
+	}
+
+	return &testNotificationService{
+		cancel: cancel,
+		ns:     ns,
+	}
+}
+
+type testNotificationService struct {
+	cancel context.CancelFunc
+
+	ns *NotificationService
+}
+
+func (tn *testNotificationService) stop() {
```

Comment:
Thinking about this you are not executing the ns Run method. So a stop is not needed. Should also add a note that you're not starting the notification service but just creating it to manually call its methods.